packages/vite-plugin-svelte/src/utils/error.js

```diff
@@ -100,3 +100,76 @@ function formatFrameForVite(frame) {
 		.map((line) => (line.match(/^\s+\^/) ? '   ' + line : ' ' + line.replace(':', ' | ')))
 		.join('\n');
 }
+
+/**
+ * @param {import('svelte/types/compiler/interfaces').Warning & Error} err a svelte compiler error, which is a mix of Warning and an error
+ * @param {string} originalCode
+ * @param {import('../index.js').Arrayable<import('svelte/types/compiler/preprocess').PreprocessorGroup>} [preprocessors]
+ */
+export function enhanceCompileError(err, originalCode, preprocessors) {
+	preprocessors = arraify(preprocessors ?? []);
+
+	/** @type {string[]} */
+	const additionalMessages = [];
+
+	// Handle incorrect TypeScript usage
+	if (err.code === 'parse-error') {
+		// Reference from Svelte: https://github.com/sveltejs/svelte/blob/800f6c076be5dd87dd4d2e9d66c59b973d54d84b/packages/svelte/src/compiler/preprocess/index.js#L262
+		const scriptRe = /<script(\s[^]*?)?(?:>([^]*?)<\/script>|\/>)/gi;
+		const errIndex = err.pos ?? -1;
+
+		let m;
+		while ((m = scriptRe.exec(originalCode))) {
+			const matchStart = m.index;
+			const matchEnd = matchStart + m[0].length;
+			const isErrorInScript = matchStart <= errIndex && errIndex <= matchEnd;
+			if (isErrorInScript) {
+				// Warn missing lang="ts"
+				if (!m[1]?.includes('lang="ts"')) {
+					additionalMessages.push('Did you forgot to add lang="ts" to your script tag?');
+				}
+				// Warn missing script preprocessor
+				if (preprocessors.every((p) => p.script == null)) {
+					additionalMessages.push(
+						'Did you forgot to add a script preprocessor? See https://github.com/sveltejs/vite-plugin-svelte/blob/main/docs/preprocess.md for more information.'
```

Comment:
this can be a bit more specific for a typescript preprocessor if we detect lang=ts is present, and else we can mention the lang value (what others are there? coffeescript? :D

Reply:
I'm focusing on TypeScript only for now as that's the main one Vite supports, and IIRC CoffeeScript doesn't really work in Vite. So TypeScript has a little special treatment here.